### ISSUE TITLE
Fatal with caching on localized fields caused by ElementDescriptor

### ISSUE THREAD

**Issue by morrisstrangfeld** [pimcore/pimcore#7367]
This is caused by the CoreHandler in the latest version. If i revert it to the state before the error is gone.
I think it relates to this:
https://github.com/pimcore/pimcore/issues/7039

![grafik](https://user-images.githubusercontent.com/30860803/96627133-b743a400-1310-11eb-8495-4ccdfd067fac.png)

**Comment by morrisstrangfeld** [comment]
FYI it only happens after sth is already cached. Clear cache resolves it temporary. It works then for around 30 sec or so until sth was added to the cache async via queue. (At least that's what i think is happening, did not debug it too deep)

**Comment by weisswurstkanone** [comment]
Unable to reproduce. Please provide more details how to create this scenario.

**Comment by morrisstrangfeld** [comment]
I tested a little deeper now and it brakes if you have blocks inside localized fields. Like so.
![grafik](https://user-images.githubusercontent.com/30860803/96714113-f968f600-13a1-11eb-8618-b26d6162f40b.png)

Caching breakes on individual objects. The delay i mentioned before was because of async caching in lists. If you call the item directly it breaks imediatelly. Happens also in the backend as soon as you add a block and save.

**Comment by weisswurstkanone** [comment]
Still not able to reproduce this scenario on https://demo.pimcore.fun/

**Comment by morrisstrangfeld** [comment]
Hmm i tried on demo with simple localized block and could not reproduce either.
I debugged on my local setup and figured our it happens while copying in the Data in CoreHandler.
![grafik](https://user-images.githubusercontent.com/30860803/96732786-05ac7d80-13b9-11eb-9692-0c8435c680df.png)
 The object of the owner switches to \Pimcore\Model\Element\ElementDescriptor after copying. Which causes the issue. But i could not figure out why. Quickfix is to switch on lazy loading for that block.
![grafik](https://user-images.githubusercontent.com/30860803/96733158-705db900-13b9-11eb-83e0-4df1c8c4887e.png)
![grafik](https://user-images.githubusercontent.com/30860803/96733208-7bb0e480-13b9-11eb-9075-13f4693134cc.png)

**Comment by morrisstrangfeld** [comment]
Hmm workaround with lazy loading does not work as well.. in the  backend there is a follow up issue now:

`URL: /admin/object/get
Message: owner missing
Trace: 
in /var/www/application/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Block.php:1226
#0 /var/www/application/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Block.php(231): Pimcore\Model\DataObject\ClassDefinition\Data\Block->setBlockElementOwner(Object(Pimcore\Model\DataObject\Data\BlockElement), Array)
#1 /var/www/application/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Block.php(1015): Pimcore\Model\DataObject\ClassDefinition\Data\Block->getDataFromResource('a:1:{i:0;a:2:{s...', Object(Pimcore\Model\DataObject\Localizedfield), Array)
#2 /var/www/application/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php(396): Pimcore\Model\DataObject\ClassDefinition\Data\Block->load(Object(Pimcore\Model\DataObject\Localizedfield), Array)
#3 /var/www/application/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php(197): Pimcore\Model\DataObject\Localizedfield->loadLazyField(Object(Pimcore\Model\DataObject\ClassDefinition\Data\Block), 'accordion_l', 'de')
#4 /var/www/application/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Localizedfields.php(179): Pimcore\Model\DataObject\Localizedfield->getInternalData(true)
#5 /var/www/application/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Localizedfields.php(141): Pimcore\Model\DataObject\ClassDefinition\Data\Localizedfields->doGetDataForEditMode(Object(Pimcore\Model\DataObject\Localizedfield), Object(AppBundle\Model\DataObject\Product), Array, Array, 1, Array)
#6 /var/www/application/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php(654): Pimcore\Model\DataObject\ClassDefinition\Data\Localizedfields->getDataForEditmode(Object(Pimcore\Model\DataObject\Localizedfield), Object(AppBundle\Model\DataObject\Product), Array)
#7 /var/www/application/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php(572): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->getDataForField(Object(AppBundle\Model\DataObject\Product), 'localizedfields', Object(Pimcore\Model\DataObject\ClassDefinition\Data\Localizedfields), false)
#8 /var/www/application/vendor/pimcore/pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php(469): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->getDataForObject(Object(AppBundle\Model\DataObject\Product), false)
#9 /var/www/application/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(158): Pimcore\Bundle\AdminBundle\Controller\Admin\DataObject\DataObjectController->getAction(Object(Symfony\Component\HttpFoundation\Request), Object(Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher))
#10 /var/www/application/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php(80): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1)
#11 /var/www/application/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php(201): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true)
#12 /var/www/application/web/app.php(36): Symfony\Component\HttpKernel\Kernel->handle(Object(Symfony\Component\HttpFoundation\Request))
#13 {main}`

**Comment by morrisstrangfeld** [comment]
@weisswurstkanone do you have any clue what could cause that?

**Comment by weisswurstkanone** [comment]
see https://github.com/pimcore/pimcore/pull/7360

**Comment by morrisstrangfeld** [comment]
Ok but this was just the follow up issue and only happens after i enabled lazy loading for the block to workaround the first issue. Because the object of the owner switched in the first place i think both issues are related.

**Comment by weisswurstkanone** [comment]
Can you please try to debug this on your local instance. As we both agreed this is not reproducable on the public demo instance.

**Comment by morrisstrangfeld** [comment]
Ok i will try to figure it out.

**Comment by morrisstrangfeld** [comment]
I updated to the latest version and the lazy load workaround works now. But the main issue still exists.
I also tried with clean pimcore installation and it happens there as well with a simple object.
I tried with and without redis.
I tried with php 7.2 7.3 and 7.4 and it happens in all of them.
Somehow it does not happen in https://demo.pimcore.fun/
Could you please try with a clean pimcore in latest version?

**Comment by morrisstrangfeld** [comment]
@weisswurstkanone i figured it out. In my testsystem i was running apache2 with mod php.
I switched to nginx with php-fpm and the issue is gone.
Really weird.

**Comment by brusch** [comment]
I'll close this for now. If there are any new findings on this please feel free to continue the discussion here. Thanks!

**Comment by morrisstrangfeld** [comment]
The finding is that it does not work with mod php and apache.
Is this intended behaviour?
Is this setup not supported?

**Comment by morrisstrangfeld** [comment]
@weisswurstkanone 
The issue persists in that specific setup. Can you explain if that behaviour is intended or if it is not supported?
I switched my local system now to nginx (had other systems anyway like that) so it is not an issue for me anymore but others might face it as well.

**Comment by wvervuurt** [comment]
I am experiencing the same error but this thread was hard to find because the error message is shown as an image instead of text. I am getting this error when running cache warming in a Homestead environment (Vagrant) running PimCore 6.8.4. If I revert to 6.7.3 everything works without a problem. 

```
bin/console pimcore:cache:warming -o object --verbose
Warming document cache for all types...
Warming asset cache for all types...
Warming object cache for types object...

In Localizedfield.php line 374:

  [Symfony\Component\Debug\Exception\UndefinedMethodException]
  Attempted to call an undefined method named "getClass" of class "Pimcore\Model\Element\ElementDescriptor".

```

**Comment by morrisstrangfeld** [comment]
I don't think it has to do with nginx or apache. It happened with mod_php (which only exists on apache) and not with php fpm (which i used with nginx).
So it seems cli might be affected as well.
@brusch would you kindly reopen the task?

**Comment by wvervuurt** [comment]
Also I don't get this problem running cache warming for _folder_ or _variant_ object types:

```
$ bin/console pimcore:cache:warming -t object -o folder
Warming object cache for types folder...
$ bin/console pimcore:cache:warming -t object -o variant
Warming object cache for types variant...
$ bin/console pimcore:cache:warming -t object -o object
Warming object cache for types object...

In Localizedfield.php line 374:

  Attempted to call an undefined method named "getClass" of class "Pimcore\Model\Element\ElementDescriptor".

```

**Comment by wvervuurt** [comment]
I can now reliably reproduce this problem in my installation.

Conditions:
- A Class definition that has a localizedfields structure
- with a Block structure in it
- with at least 1 value field in it (I tested with a single Text Input field and with a single Textarea)
- and at least 1 object instance of that Class, that has at least 1 value in that Block

In this case I get the error. If I remove the single value from the object instance's Block, everything works fine, so the problem doesn't seem to be the class structure itself but the existence of objects with actual values in them.

When I had a problematic object instance and changed the class structure by moving the Block outside of the localizedfields structure, the problem disappeared (as expected based on the file in which the error occurs).

**Comment by wvervuurt** [comment]
I have debugged some more in an empty PimCore installation with only 1 class and 1 instance.

While running the _pimcore:cache:warming_ command in CLI I noticed that the LocalizedField::getFieldDefinitions method is being called 4 times for this single instance object. The first 3 times are all OK, the fourth time is when $this->getObject() (at line 374 in Localizedfield.php) returns a Pimcore\Model\Element\ElementDescriptor object instead of the expected DataObject.

When I check the stack trace in the non-problematic cases, I get this:
```
  at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:374
 Pimcore\Model\DataObject\Localizedfield->getFieldDefinitions() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:784
 Pimcore\Model\DataObject\Localizedfield->getLazyLoadedFieldNames() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:188
 Pimcore\Model\DataObject\Localizedfield->getInternalData() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/ClassDefinition/Data/Localizedfields.php:956
 Pimcore\Model\DataObject\ClassDefinition\Data\Localizedfields->getCacheTags() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Concrete.php:371
 Pimcore\Model\DataObject\Concrete->getCacheTags() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Core/CoreHandler.php:498
 Pimcore\Cache\Core\CoreHandler->prepareCacheTags() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Core/CoreHandler.php:370
 Pimcore\Cache\Core\CoreHandler->save() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache.php:95
 Pimcore\Cache::save() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:102
 Pimcore\Cache\Tool\Warming::loadElementToCache() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:123
 Pimcore\Cache\Tool\Warming::loadToCache() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:74
 Pimcore\Cache\Tool\Warming::objects() at /home/vagrant/code/vendor/pimcore/pimcore/bundles/CoreBundle/Command/CacheWarmingCommand.php:144
 Pimcore\Bundle\CoreBundle\Command\CacheWarmingCommand->execute() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:1027
 Symfony\Component\Console\Application->doRunCommand() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:273
 Symfony\Component\Console\Application->doRun() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:149
 Symfony\Component\Console\Application->run() at /home/vagrant/code/bin/console:26
```

In the problematic case, however, the calls to Localizedfield::getFieldDefinitions come from somewhere else:
```
  at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:374
 Pimcore\Model\DataObject\Localizedfield->getFieldDefinitions() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:784
 Pimcore\Model\DataObject\Localizedfield->getLazyLoadedFieldNames() at /home/vagrant/code/vendor/pimcore/pimcore/models/DataObject/Localizedfield.php:659
 Pimcore\Model\DataObject\Localizedfield->__sleep() at n/a:n/a
 serialize() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Core/CoreHandler.php:616
 Pimcore\Cache\Core\CoreHandler->storeCacheItem() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Core/CoreHandler.php:375
 Pimcore\Cache\Core\CoreHandler->save() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache.php:95
 Pimcore\Cache::save() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:102
 Pimcore\Cache\Tool\Warming::loadElementToCache() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:123
 Pimcore\Cache\Tool\Warming::loadToCache() at /home/vagrant/code/vendor/pimcore/pimcore/lib/Cache/Tool/Warming.php:74
 Pimcore\Cache\Tool\Warming::objects() at /home/vagrant/code/vendor/pimcore/pimcore/bundles/CoreBundle/Command/CacheWarmingCommand.php:144
 Pimcore\Bundle\CoreBundle\Command\CacheWarmingCommand->execute() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:1027
 Symfony\Component\Console\Application->doRunCommand() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:273
 Symfony\Component\Console\Application->doRun() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /home/vagrant/code/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:149
 Symfony\Component\Console\Application->run() at /home/vagrant/code/bin/console:26
```

So it looks like something is going wrong that has something to do with lazy loading. I'm not too familiar with that side of PimCore but I do recall having seen more issues w.r.t. problems caused by lazy loading and Blocks in Localizedfields.

I also have a hunch that this is somehow related to #7060 but I'm entirely unsure.

@brusch and @weisswurstkanone, does this info help at all to pinpoint the cause of all this?

**Comment by wdelfuego** [comment]
A small change that makes the error not occur is changing the check on line 627 of Localizedfield.php from

`        if (!$this->isInDumpState()) {`

to

`        if ($this->getObject() instanceof \Pimcore\Model\DataObject && !$this->isInDumpState()) {`

I have no idea if this has unintended side effects so I'm not sure it is even a valid workaround, just wanted to share this bit of info.

**Comment by wdelfuego** [comment]
Thanks for reopening, @brusch !

Would you say the added check for _instanceof \Pimcore\Model\DataObject_ as proposed in my previous comment is a safe work-around for this issue?

**Comment by wdelfuego** [comment]
Awesome! Thanks for the fast fix, @weisswurstkanone. Tested and confirmed to work 👍

**Comment by morrisstrangfeld** [comment]
But really weird that it does not happen with php-fpm.
Would be really intersted in the cause of this.
Do you have any details @weisswurstkanone ?

**Comment by weisswurstkanone** [comment]
I guess this heavily depends on the setup and current working environment. 

It feeled to be something random but it isn't that at all. It was quite easy to reproduce with your valuable insights!

Thanks a lot for your contributions!

**Comment by dvesh3** [comment]
Fixed by #7513
